MigrateServicePreferenceFromLegacy/handler.ts

```diff
@@ -0,0 +1,139 @@
+} from "@pagopa/io-functions-commons/dist/src/models/service_preference";
+import {
+  CosmosErrorResponse,
+  CosmosErrors
+} from "@pagopa/io-functions-commons/dist/src/utils/cosmosdb_model";
+import { NonNegativeInteger } from "@pagopa/io-functions-commons/node_modules/@pagopa/ts-commons/lib/numbers";
+import { ResponseErrorFromValidationErrors } from "@pagopa/io-functions-commons/node_modules/@pagopa/ts-commons/lib/responses";
+import * as a from "fp-ts/lib/Array";
+import * as e from "fp-ts/lib/Either";
+import * as o from "fp-ts/lib/Option";
+import * as te from "fp-ts/lib/TaskEither";
+import * as t from "io-ts";
+import { FiscalCode } from "../generated/backend/FiscalCode";
+import { ServiceId } from "../generated/backend/ServiceId";
+
+const COSMOS_ERROR_KIND = "COSMOS_ERROR_RESPONSE";
+const CONFLICT_CODE = 409;
+const LOG_PREFIX = "MigrateServicePreferenceFromLegacy";
+
+export const MigrateServicesPreferencesQueueMessage = t.interface({
+  newProfile: RetrievedProfile,
+  oldProfile: RetrievedProfile
+});
+export type MigrateServicesPreferencesQueueMessage = t.TypeOf<
+  typeof MigrateServicesPreferencesQueueMessage
+>;
+
+function isCosmosError(
+  ce: CosmosErrors
+): ce is ReturnType<typeof CosmosErrorResponse> {
+  return ce.kind === COSMOS_ERROR_KIND;
+}
+
+export const createServicePreference = (
+  serviceId: ServiceId,
+  blockedChannels: ReadonlyArray<BlockedInboxOrChannelEnum>,
+  fiscalCode: FiscalCode,
+  version: NonNegativeInteger
+): NewServicePreference => ({
+  fiscalCode,
+  id: makeServicesPreferencesDocumentId(fiscalCode, serviceId, version),
+  isEmailEnabled: !blockedChannels.some(
+    channel => channel === BlockedInboxOrChannelEnum.EMAIL
+  ),
+  isInboxEnabled: !blockedChannels.some(
+    channel => channel === BlockedInboxOrChannelEnum.INBOX
+  ),
+  isWebhookEnabled: !blockedChannels.some(
+    channel => channel === BlockedInboxOrChannelEnum.WEBHOOK
+  ),
+  kind: "INewServicePreference",
+  serviceId,
+  settingsVersion: version
+});
+
+export const blockedsToServicesPreferences = (
+  blocked: {
+    [x: string]: readonly BlockedInboxOrChannelEnum[];
+  },
+  fiscalCode: FiscalCode,
+  version: NonNegativeInteger
+) =>
+  o
+    .fromNullable(blocked)
+    .map(b =>
+      Object.entries(b)
+        .filter(i => ServiceId.is(i[0]))
+        .map(
+          i =>
+            createServicePreference(
+              i[0] as ServiceId,
+              i[1],
+              fiscalCode,
+              version
+            ) // cast required: ts do not identify filter as a guard
+        )
+    )
+    .getOrElse([]);
+
+export const MigrateServicePreferenceFromLegacy = (
+  servicePreferenceModel: ServicesPreferencesModel
+) => async (context: Context, input: unknown) =>
+  te
+    .fromEither(MigrateServicesPreferencesQueueMessage.decode(input))
+    .mapLeft(
+      ResponseErrorFromValidationErrors(MigrateServicesPreferencesQueueMessage)
+    )
+    .mapLeft(err => new Error(err.detail))
```

Comment:
why double `mapLeft` ?

Reply:
to collect the errors from validation in a single generic "Error" without casting. 
Is there a ResponseErrorFromValidationErrors method "version" doing it?